### PR TITLE
Added `recommendItemsForUserItemSubset` method to spark `ALS`

### PR DESCRIPTION
**Pull request by netang**
Scala:
* copied `org.apache.spark.ml.recommendation` package (from [ALS.scala](https://github.com/apache/spark/blob/branch-3.1/mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala)) to `org.apache.spark.ml.recommendation.replay` package.
* added `recommendItemsForUserItemSubset` method to `org.apache.spark.ml.recommendation.replay` package.

Python:
* copied `pyspark.ml.recommendation` package to `replay.spark_custom_models.recommendation` package.
* added `recommendItemsForUserItemSubset` method and changed `ALS.__init__` method in `replay.spark_custom_models.recommendation`.

### CODE REVIEW

File: scala/src/main/scala/org/apache/spark/ml/recommendation/replay/ReplayALS.scala [hunk 1]
```diff
@@ -0,0 +1,1824 @@
+/*
```
Comment (monkey0head) [inline scala/src/main/scala/org/apache/spark/ml/recommendation/replay/ReplayALS.scala:1]: we need to assure coping and rewriting the code is in accordance with the license,

File: replay/spark_custom_models/recommendation.py [hunk 2]
```diff
@@ -0,0 +1,721 @@
+#
```
Comment (monkey0head) [inline replay/spark_custom_models/recommendation.py:1]: what about the license?

**Comment by monkey0head** [comment]
please list the methods to review from replay/spark_custom_models/recommendation.py and .scala

File: scala/src/main/scala/org/apache/spark/ml/recommendation/replay/ReplayALS.scala [hunk 3]
```diff
@@ -0,0 +1,1824 @@
...
+   *         stored as an array of (userCol: Int, rating: Float) Rows.
+   */
+  @Since("2.3.0")
+  def recommendForItemSubset(dataset: Dataset[_], numUsers: Int): DataFrame = {
+    val srcFactorSubset = getSourceFactorSubset(dataset, itemFactors, $(itemCol))
+    recommendForAll(srcFactorSubset, userFactors, $(itemCol), $(userCol), numUsers, $(blockSize))
+  }
+
+  @Since("2.3.0")
+  def recommendItemsForUserItemSubset(usersDataset: Dataset[_], itemsDataset: Dataset[_], numItems: Int): DataFrame = {
```
Comment (netang) [inline scala/src/main/scala/org/apache/spark/ml/recommendation/replay/ReplayALS.scala:413]: @monkey0head

File: scala/src/main/scala/org/apache/spark/ml/recommendation/replay/ReplayALS.scala (hunk 1 above)
Comment (monkey0head) [reply scala/src/main/scala/org/apache/spark/ml/recommendation/replay/ReplayALS.scala:1]: we do not have any NOTICE file in our repo, so the line 3 below becomes wrong.

have a look at section 4, which contains info about the NOTICE file:
https://www.apache.org/licenses/LICENSE-2.0

spark notice file is here: 
https://github.com/apache/spark/blob/master/NOTICE

could you bring your work up to the standard?

**Comment by monkey0head** [comment]
please resolve the conflict

**Review by monkey0head — APPROVED** [review]

**Comment by monkey0head** [comment]
will be merged with https://github.com/sb-ai-lab/RePlay/pull/47

**Pull request closed without merging**